operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/Controller.java

```diff
@@ -58,40 +52,42 @@ public Controller(Reconciler<P> reconciler,
 
   @Override
   public DeleteControl cleanup(P resource, Context<P> context) {
-    dependents.cleanup(resource, context);
-
     try {
-      return metrics().timeControllerExecution(
-          new ControllerExecution<>() {
-            @Override
-            public String name() {
-              return "cleanup";
-            }
-
-            @Override
-            public String controllerName() {
-              return configuration.getName();
-            }
-
-            @Override
-            public String successTypeName(DeleteControl deleteControl) {
-              return deleteControl.isRemoveFinalizer() ? "delete" : "finalizerNotRemoved";
-            }
-
-            @Override
-            public DeleteControl execute() {
-              return reconciler.cleanup(resource, context);
-            }
-          });
+      return metrics()
+          .timeControllerExecution(
+              new ControllerExecution<>() {
+                @Override
+                public String name() {
+                  return "cleanup";
+                }
+
+                @Override
+                public String controllerName() {
+                  return configuration.getName();
+                }
+
+                @Override
+                public String successTypeName(DeleteControl deleteControl) {
+                  return deleteControl.isRemoveFinalizer() ? "delete" : "finalizerNotRemoved";
+                }
+
+                @Override
+                public DeleteControl execute() {
+                  dependents.cleanup(resource, context);
```

Comment:
I agree, dependent cleanup should be called only if the reconciler implement `Cleaner`. Maybe a warning or a controller configuration exception should be raised if dependents and reconciler do not align ?

Reply:
I think this is kinda a side effect that we don't have this in layers. In case of a special Reconciler this could be done very transparent way. Will create one prototype in the future.

If you take a look from the other point of view, event the `reconcile` method could be empty when using managed dependent resources, what is already weird. So if we have managed dependent resources where does implement cleaner, the custom resources must have a finalizer otherwise it does not work. And also `cleanup` needs to be called.

I agree that this is little weird, but the other alternative is to just log a warning, and don't do finalizer and cleanup call, is very error prone in sense that could lead easily to operator implementation that is already not correct.

So personally I would go this way, so rather have under the hood something happening that is correct but maybe not intuitive (however nicely documented), and users could then rather check learn why is this happening. It might cause less headaches then having not cleaned up resources, because `Cleaner` is not implement on reconciler.

Reply:
But I don't have actually that strong opinion, just lean to this. Let me know what do you think about this reasoning, we can change in case.